tests/unit/lms/db/_bulk_action_test.py

```diff
@@ -58,6 +77,60 @@ def test_upsert(self, db_session):
     def test_upsert_does_nothing_if_given_an_empty_list_of_values(self, db_session):
         assert BulkAction(db_session).upsert(self.TableWithBulkUpsert, []) == []
 
+    def test_upsert_with_onupdate_columns(self, db_session):
+        db_session.add_all(
+            [
+                self.TableWithBulkUpsertOnUpdate(
+                    id=1, name="existing", scalar=0, callable=0, sql=0, default=0
+                ),
```

Comment:
I was having a little trouble working out what the focus of this test is with the different columns, until I realised that this is testing all of the different variants at once.

It might help reveal the focus of the test and make it a little shorter as well to parameterise the columns so they are tested one by one:

```suggestion
    @pytest.mark.parametrize("column", ("scalar", "callable", "sql", "default"))
    def test_upsert_with_onupdate_columns(self, db_session, column):
        db_session.add_all(
            [
                self.TableWithBulkUpsertOnUpdate(id=1, name="pre_existing_1", **{column: 0}),
                ...
```